### PR TITLE
feat: Derive

### PR DESCRIPTION
**Pull request by hellerve**
This PR adds `derive`, a macro for deriving arbitrary interfaces for structs based on their members. It is also pluggable, meaning that you can write your own derivers.

This PR is a draft, because it depends on #1101 being fixed ~~and because documentation has to be written on the feature~~.

Cheers

EDIT: Documentation has been written.

### CODE REVIEW

**Comment by hellerve** [comment]
For documentation purposes I wanted to put `derive` into a module, but it seems like #1121 prevents me from doing so.

**Comment by hellerve** [comment]
This is ready for review!

File: docs/Derive.md
Comment (scolsen) [inline docs/Derive.md:9]: Maybe link to the sections here using heading anchors?` [first section](#i-using-derive)`

File: core/Derive.carp
Comment (scolsen) [inline core/Derive.carp:61]: I don't think it's totally necessary, but I feel this might read a bit better if `t` is the second argument:

`(derive zero Point)`

This also would make it follow the order of `implements`

`(implements zero foo)`

File: docs/Derive.md
Comment (scolsen) [inline docs/Derive.md:26]: How about:

"The code above will provide implementations of `zero` and `=` for the type `Point` based on its members."

File: docs/Derive.md
Comment (scolsen) [inline docs/Derive.md:47]: Link to section 2

File: docs/Derive.md
Comment (scolsen) [inline docs/Derive.md:48]: I would delete "you with", so "Carp provides a special dynamic..."

**Review by scolsen — APPROVED** [review]
Amazing! Such a useful feature.

I left a few comments, mostly nits.

File: core/Derive.carp
Comment (hellerve) [reply core/Derive.carp:20]: That makes sense! Resolved in `0b8ae4b6`.

File: core/Derive.carp
Comment (hellerve) [reply core/Derive.carp]: Fair enough; what’s the consensus here, @carp-lang/maintainers?

**Comment by hellerve** [comment]
Thanks for the edits, @scolsen! I implemented them (except for the order of arguments in `derive`, which I’m waiting for feedback on).

File: core/Derive.carp
Comment (eriksvedang) [reply core/Derive.carp]: I think it makes a lot of sense to have the same order for derive & implements. Since we already have a lot of code that relies on the `implements` ordering it seems reasonable to go with that.

Comment (TimDeve) [reply core/Derive.carp]: Always felt that the order in `implements` should be the other way around, especially if you read it "infix": `zero implements foo`. But there is certainly something to be said for consistency.

Comment (scolsen) [reply core/Derive.carp]: > Always felt that the order in implements should be the other way around, especially if you read it "infix": zero implements foo.
But there is certainly something to be said for consistency.

That's a fair point, if others feel this way too, I'm ok with making a change on the `implements` side. The reason it has the order is because when considering prefix notation, `(implements zero foo)` made more sense to me--basically imagine there's a comma or colon when reading it off as english-- `implements zero: foo`--it also allows the form to take varargs in the future `(implements zero foo...)`. 

`(implements foo zero)`, putting the implementation first, felt odd and seemed more likely to create friction when skimming code even if it corresponds to the correct infix order. At the very least, I'd personally assume `foo` was the interface in that case, why precisely I don't know but it's how my brain wants to naturally parse it (this may well be a personal problem, so I'm happy to go along with consensus).

Another argument for changing the order on the `implements` side would be that it's more consistent, I think, with some of our other forms, `meta` for example.

Comment (eriksvedang) [reply core/Derive.carp]: It's definitely not super clear which one is most natural. I usually don't try to make it "read naturally" in the natural language kind of way. Instead (even when now programming Haskell) I use some kind of "currying" mindset where the bigger or more stable argument comes first to the function. This is of course kind of vague, but most obvious example is maybe a setter; the thing being modified comes first and then comes the new value.

Applying the same idea to the functions we're discussing seem to suggest that `implements` should switch, since we're "modifying that type". One could maybe also say that we're modifying the interface, but I don't think that's how people generally think about it (even though it's kind of what's happening in the compiler internals.)

All this is to say that I kind of agree that this PR has the superior order, but it's a close call and I didn't want to suggest changing all the other code in the same PR. 

Maybe we can just merge this as-is and fix the ordering later? That will give us more time to think about it, without blocking people from trying out `derive`.

Comment (hellerve) [reply core/Derive.carp]: I see merits in either ordering, and I don’t mind either. What would be interesting would be to enable people to derive multiple interfaces in one call:

```clojure
(derive [= zero str] T)
; vs
(derive T [= zero str])
```

Where, for me, the second ordering seems more natural; this of course is entirely subjective as well.

I’d argue that for now keeping concistency and not breaking existent behavior probably trumps any need to change, meaning that I should go ahead and switch the order for `derive` and we can revisit this another time. How does that sound?

Comment (eriksvedang) [reply core/Derive.carp]: Deriving multiple ones sounds great. Could be a `*` macro also, of course (or just rest args?)

> How does that sound?

Perfect!

Comment (hellerve) [reply core/Derive.carp]: While changing the order I realized that this would also change overrides to look like this:

```clojure
(derive = T eq) ; => derive = for T as eq
```

I don’t know whether I like that very much.

Comment (eriksvedang) [reply core/Derive.carp]: ugh, no that's really confusing

Comment (TimDeve) [reply core/Derive.carp]: `(derive T [= zero str])` Cool 🙂

Comment (scolsen) [reply core/Derive.carp]: Cool, good discussion! Let's keep this order and change the order of `implements` instead. I can take that on.

Comment (hellerve) [reply core/Derive.carp]: Okay, I changed the order back for now then :) I’ll resolve this discussion for now!

File: docs/Derive.md
Comment (eriksvedang) [inline docs/Derive.md:31]: typo: "wil"

**Review by eriksvedang — COMMENTED** [review]
Found a typo.

**Comment by scolsen** [comment]
Thanks @hellerve sorry for all the back and forth

**Comment by hellerve** [comment]
No worries, that’s how design discussion (should) go!

**Comment by eriksvedang** [comment]
Is this ready to merge now? It looks great to me, but there was so much discussion that I might have missed something.

**Comment by hellerve** [comment]
No objections from me, I think we are ready! 🚀

**Pull request merged**
